Initial Classes

mecha opened the pull request:
The initial abstract classes and traits, together with default generic implementations: `Menu` and `MenuItem`.

mecha committed:
Added basic-most traits

mecha committed:
Added `AbstractMenuElement`

mecha committed:
Added `AbstractMenu` and `AbstractMenuItem`

mecha committed:
Added `AbstractBaseMenuElement`

mecha committed:
Added `AbstractBaseMenu`

mecha committed:
Added `AbstractBaseMenuItem`

mecha committed:
Added `CallbackAwareTrait`

mecha committed:
Added `Menu` and `MenuItem` classes

mecha commented:
Unit tests for `Menu` and `MenuItem` are currently incomplete.

mecha commented:
A note on the hierarchy, which can be confusing since it attempts to keep both menus and menu items using a common ancestor.

All abstract functionality can be found in `MenuElementTrait`, `MenuTrait` and `MenuItemTrait`. There are also three abstract classes, each using one of these three traits, but only the `AbstractMenuElement` class is used in this package, which `AbstractBaseMenuElement` extends.

`AbstractBaseMenu` and `AbstractBaseMenuItem` then extend that, which are used to create the concrete implementations.

----

This is no longer true since recent refactoring. Now there is no distinction between a menu and a menu item. All menu nodes are `MenuElementInterface` instances, `ChildrenAwareNodeInterface` and `KeyAwareNodeInterface`.

mecha committed:
`Menu` now implements `WriteableMenuInterface`

mecha committed:
Updated dependencies

mecha committed:
Added unit test for `Menu`

mecha committed:
Fixed use statements in `MenuTest`

mecha committed:
Added unit test for `MenuItem`

mecha committed:
Updated dependencies

mecha committed:
Removed old unused classes

mecha committed:
Outsourced functionlity in `MenuElementTrait`

mecha committed:
Added `IconAwareTrait`

mecha committed:
Removed `AbstractMenuElement#_onSelected()`

mecha committed:
`AbstractBaseMenuElement` now properly implements ...

... `MenuElementInterface` and `NodeInterface` methods

mecha committed:
`AbstractMenuElement` validates parent and children

mecha committed:
Added `CapabilityAwareTrait`

mecha committed:
Removed old functional tests

mecha committed:
Fixed functional test for `AbstractMenuElement`

mecha committed:
Fixed trait functional tests

mecha committed:
Removed redundant tests from `MenuElementTraitTest`

mecha committed:
Fixed functional test for `AbstractBaseMenuElement`

mecha committed:
Fixed imports in `AbstractBaseMenuElement`

mecha committed:
Added explaination in `MenuElementTraitTest`

mecha committed:
Fixed imports in `MenuElementTraitTest`

mecha committed:
`AbstractMenuElement` handles child positioning

mecha committed:
Updated repository URLs

mecha committed:
Added `prefer-stable` flag

mecha committed:
Updated dependencies

mecha committed:
Updated dependencies

mecha committed:
Added `CallbackMenu` and test

mecha committed:
Menu elements are now child and key aware nodes

Also removes the `MenuElementTrait`. The `AbstractMenuElement` is now the only functional foundation class.

mecha committed:
Fixed code standards

Signed-off-by: Miguel Muscat <miguelmuscat93@gmail.com>

mecha committed:
Updated dependencies

mecha committed:
Updated dependencies

mecha committed:
Removed `CallbackMenu` child menu registration

mecha committed:
Fixed unit test for `CallbackMenu`

mecha committed:
`AbstractMenuElement` initializes children array

mecha committed:
`AbstractBaseMenuElement` is now a `RecursiveIterator`

mecha committed:
Fixed test for `AbstractBaseMenuElement`

mecha committed:
Added iteration test for `AbstractBaseMenuElement`

mecha committed:
Added constants for built in sidebar menus

mecha committed:
Added `AbstractMenuBar`

mecha committed:
Fixed namespace in `CallbackMenu` unit test

mecha committed:
Added `UrlMenu` and test

mecha committed:
Added `Sidebar`

mecha committed:
Added `Toolbar`

mecha committed:
Added WP Mock lib dependency

mecha committed:
Test bootstrap initializes WP Mock

mecha committed:
Added missing doc and use in `AbstactMenuBar` test

mecha committed:
Added unit test for `Toolbar`

mecha committed:
Removed use of unused property in `Sidebar`

mecha committed:
Added missing return in `Sidebar#_registerSubMenu`

mecha committed:
Removed `MenuElementInterface#getParent()` mock method

mecha committed:
Fixed position calculation in `Sidebar`

mecha committed:
Added unit test for `Sidebar`

mecha committed:
Added `PageMenu` and test

mecha committed:
Removed empty return in `Sidebar`.

XedinUnknown commented on src/AbstractMenuElement.php:
```diff
@@ -0,0 +1,213 @@
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
```
Looks like `@throws` is missing

mecha replied:
It's inherited from `ChildrenAwareTrait#_validateChild()`.

XedinUnknown commented on src/AbstractMenuElement.php:
```diff
@@ -0,0 +1,213 @@
+     * @param int|null $position An integer, with larger values signifying a lower position, or null to append.
+     *
+     * @return $this
+     */
```
Looks like `@throws` is missing

XedinUnknown commented on src/AbstractMenuElement.php:
```diff
@@ -0,0 +1,213 @@
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
```
Looks like `@throws` is missing.

mecha replied:
Method has been removed. Reason: menu elements should not be parent-aware.

XedinUnknown commented on src/AbstractMenuElement.php:
```diff
@@ -0,0 +1,213 @@
+     * @param int                                         $code             The exception code.
+     * @param null                                        $inner            The previous exception in the chain.
+     * @param mixed                                       $subject          The subject that failed validation.
+     * @param string[]|StringableInterface[]|\Traversable $validationErrors The validation errors.
```
Could we make this a bit shorter?

mecha replied:
Don't know. Can we?

XedinUnknown replied:
E.g. by aliasing `StringableInterface` to `Stringable`, like I always do, and by `use`ing `Traversable`.

XedinUnknown commented on src/AbstractMenuElement.php:
```diff
@@ -0,0 +1,213 @@
+    protected function _validateParent($parent)
+    {
+        if ($parent !== null && !$this->_isMenuElement($parent)) {
+            throw $this->_createValidationFailedException('Parent is not a valid menu element.', 0, null, $parent);
```
Looks like no reasons for failed validation are given.

mecha replied:
Actually, menus are not meant to be parent-aware. I had forgotten to remove the trait.

XedinUnknown commented on src/AbstractMenuElement.php:
```diff
@@ -0,0 +1,213 @@
+    protected function _validateChild($child)
+    {
+        if (!$this->_isMenuElement($child)) {
+            throw $this->_createValidationFailedException('Child is not a valid menu element.', 0, null, $child);
```
Looks like no reasons for failed validation are given.

XedinUnknown commented on src/AbstractMenuElement.php:
```diff
@@ -0,0 +1,213 @@
+
+    /**
+     * Normalizes the position.
+     *
```
A description of what happens would be great.

XedinUnknown commented on src/AbstractBaseMenuElement.php:
```diff
@@ -0,0 +1,200 @@
+     */
+    public function getChildren()
+    {
+        return new RecursiveArrayIterator($this->_getChildren());
```
~Make factory method for this.~ Let's talk about recursive iteration first.

XedinUnknown commented on src/AbstractBaseMenuElement.php:
```diff
@@ -0,0 +1,200 @@
+     */
+    public function hasChildren()
+    {
+        // What to do here? ChildrenAware::hasChildren() or RecursiveIterator::hasChildren() ?
```
Yes, we have a problem. I also faced this problem with my `ConfigInterface` implementation. I have an idea of how to solve this using proper SoC. Let's talk about that ASAP.

XedinUnknown reviewed (changes requested):
Documentation needs to be added about what kind of exceptions methods throw.
Also, the recursive iteration seems to be a problem to discuss in-depth.

mecha committed:
Set NB PHP version to 5.4

mecha committed:
Menu elements are no longer parent aware

mecha committed:
Documented `AbstractMenuElement#_normalizePosition()`

mecha committed:
Improved `AbstractMenuElement#_normalizePosition()` to ...

... be more readable.

mecha committed:
Added factory method for children iterator

mecha committed:
Removed 5.3 from Travis CI builds

mecha committed:
Added iterator package dependencies

mecha committed:
Added `RecursiveIteration`

mecha committed:
Added `RecursiveChildrenAwareIterator`

mecha committed:
`AbstractBaseElement` is an `IteratorAggregate` that ...

... delegates to a `RecursiveChildrenAwareIterator`

mecha committed:
Fixed code standards

mecha committed:
Outsourced recursive iteration to `dhii/iterator-base`

mecha committed:
Added test for `RecursiveMenuIterator`

mecha committed:
Removed unused use statement

mecha committed:
Fixed code standards

XedinUnknown commented:
OMG the `class RecursiveMenuIterator extends AbstractBaseRecursiveIterator` is so amaze!!! @mecha, I cri of joi.

XedinUnknown committed:
Attempting workaround for failed Travis builds

See travis-ci/travis-ci#8113

XedinUnknown committed:
Another `no-api` flag added

Trying to fix dep `rebelcode/wp-admin-page-interface`. The previous fix seems to have worked.

XedinUnknown committed:
Changed build distro to `precise`

This could fix failing builds.

mecha committed:
Updated version constraints

mecha committed:
Merge branch 'task/initial-classes' of https://github.com/RebelCode/wp-admin-menu into task/initial-classes

mecha committed:
Updated admin page lib version

XedinUnknown committed:
Fixed project dependencies

XedinUnknown committed:
Now depending on `url-interface`

XedinUnknown committed:
Fixed implementation and tests

XedinUnknown committed:
Merge branch 'task/initial-classes' of https://github.com/RebelCode/wp-admin-menu into task/initial-classes

Conflicts:
	composer.json
	composer.lock

XedinUnknown committed:
Added PHPS project files

XedinUnknown committed:
Now also building on PHP 7.2

XedinUnknown committed:
Now allowing higher versions of `wp_mock`

Some tests will be run only on higher versions.
